### PR TITLE
Review adjustments to `smartquery` and `query_factory`

### PR DESCRIPTION
**Pull request by peekjef72**
### About
I've tried to use query methods from grafana-client info grafana-snapshot-tool. For that I need to perform some adjustement to api interface.

- extend query_factory to be able to set more parameters for each "known" datasource according to their parameters. Goal it to use parameters set in the panel mixed with datasource requirements and format to generate a request object (method, uri fragments,  data to Post)
- review examples datasource-smartquery.py and datasource-health-check.py to ensure they still work.

### Discussion
it seems to me that :
- datasource accesses change with grafana version: Prometheus in v 7 use format specified in function query() and query_range(), but now use a standard format for the queries (type or range) and results.
- the format of the results changes from one datasource to anothers, maybe for grafana vesion too.
- some of your tests (Loki ) seems not working for me : format my have change... not sure.

I'm not sure that the grafana_client must implement a generic function smartquery() that permit to users to query and version to Grafana server and any data source in a unified way.
What is your opinion ?

### CODE REVIEW

**Comment by peekjef72** [comment]
I've tried to use query methods from grafana-client info grafana-snapshot-tool.
For that I need to perform some adjustement to api interface:
- extend query_factory to be able to set more parameters for each "known" datasource according to their parameters. Goal it to use parameters set in the panel mixed with datasource requirements and format to generate a request object (method, uri fragments,  data to Post)
- review examples datasource-smartquery.py and datasource-health-check.py to ensure they still work.


Discussion:
it seems to me that :
- datasource accesses change with grafana version: Prometheus in v 7 use format specified in function query() and query_range(), but now use a standard format for the queries (type or range) and results.
- the format of the results changes from one datasource to anothers, maybe for grafana vesion too.
- some of your tests (Loki ) seems not working for me : format my have change... not sure.

I'm not sure that the grafana_client must implement a generic function smartquery() that permit to users to query and version to Grafana server and any data source in a unified way.
What is your opinion ?

**Comment by amotl** [comment]
Dear Jean-Francois,

thanks a stack for your excellent contribution and for bringing up the corresponding discussion topics.

> I've tried to use query methods from `grafana-client` in `grafana-snapshot-tool`. For that I need to perform some adjustments to the API interface.

`grafana-client` should definitively be able to support `grafana-snapshot-tool` in the way you are proposing. It is very sweet to see you have been able to build upon my recent additions at `grafana_client.elements.datasource`, where you certainly improved some important details in order to make it more generic. This is the place where I had to stop on the last iteration, so I am exceptionally happy about it.

I will try to dedicate some time for a detailed review of your patch soon. I think your recent updates, especially removing `smartquery_old`, makes the overall appearance much more clear now.

With kind regards,
Andreas.

File: grafana_client/elements/datasource.py [hunk 1]
```diff
@@ -301,11 +312,26 @@ def smartquery(self, datasource: Union[DatasourceIdentifier, Dict], expression:
...
+                    request["data"]["to"],
+                )
+            else:
+                return self.query_range(
+                    datasource.get("id"),
+                    request["expr"],
+                    request["data"]["from"],
+                    request["data"]["to"],
+                    request["data"]["step"],
+                )
```
Comment (amotl) [inline grafana_client/elements/datasource.py:330]: > datasource accesses change with grafana version: Prometheus in v 7 use format specified in function query() and query_range(), but now use a standard format for the queries (type or range) and results.

Hi. It looks like @PankajKhanwani is looking for this improvement around properly supporting Prometheus at GH-85, right?

Comment (peekjef72) [reply grafana_client/elements/datasource.py:330]: Hi,
I think so.
even if PR was built for grafana 9.2 I think that the Prometheus datasource has not changed.
Do you believe you have time to review ?
JFPIK-

**Comment by amotl** [comment]
Dear Jean-Francois,

having to apologize again for the long delay, I finally found some time to review your patch in detail. I really like it, and would like to wrap it up, to satisfy the linter and the software tests, in order to finally integrate it. I would start pushing directly to your feature branch, after rebasing it, and will also squash the commits afterwards. Do you have any objections about this?

> datasource accesses change with grafana version: Prometheus in v 7 use format specified in function query() and query_range(), but now use a standard format for the queries (type or range) and results.

It is sad, right. I was well aware of those details when starting this experiment.

> the format of the results changes from one datasource to anothers, maybe for grafana vesion too.

C'est la vie, I would say. Thank you so much for making the best out of it, by introducing corresponding version switching.

> some of your tests (Loki ) seems not working for me : format my have change... not sure.

Thanks, I will look into it when starting to work on this.

> I'm not sure that the grafana_client must implement a generic function smartquery() that permit to users to query and version to Grafana server and any data source in a unified way. What is your opinion?

I am absolute not sure about it, but I wanted give it a try, because I thought it would make a nice feature. I know there will always be anomalies wrt. to a completely unified interface, that is natural. But still, I think the outcome is quite nice, after learning a bit of query parameters, very much usable for scripting purposes for automating Grafana. At least, this is what I was originally humbly aiming at.

With kind regards,
Andreas.

**Comment by amotl** [comment]
Closing this in favor of GH-112, where I added some adjustments to make the tests work again. Your improvements will be included into the next release, and I will be happy to receive corresponding feedback or bugfix patches, in case I made any mistakes. Thank you again, for both your patch and your patience.

**Pull request closed without merging**